### ISSUE TITLE
Stopped connected session, due to unknown qos2 pubrec

### ISSUE THREAD

**Issue by srados** [vernemq/vernemq#1312]
### Environment

- VerneMQ Version: 1.9.1
- OS: Linux
- Erlang/OTP version (if building from source):
- VerneMQ configuration (vernemq.conf) or the changes from the default
- Cluster size/standalone: standalone

### Expected behavior

- The server should not close the session due to unknown pubrec xxxx

### Actual behaviour
client is disconnected. the message in vernemq log is:
stopped connected session, due to unknown qos2 pubrec xxxx

steps to reproduce:
using paho go library, make a mqtt client, then,
1. use file store for mqtt client persistence 
2. on message receive, make mqtt client exit before it manages to sent the pubrec
3. the i.123.msg remains on disk
4. start the mqtt client again
5. the client gets disconnected with the message in the vernemq log being: `stopped connected session, due to unknown qos2 pubrec 123`

I understand that it may be that the mqtt client tried to send duplicate pubrec, but I think that vernemq should simply ignore it in that case.

**Comment by larshesel** [comment]
Hi, thanks for opening the issue. We'll take a look at it soon.

**Comment by srados** [comment]
btw guys at mosquitto had a similar problem https://github.com/eclipse/mosquitto/issues/57

**Comment by larshesel** [comment]
I looked at our test-suites and we have a case covering this scenario where we disconnect after having received the QoS 2 PUBLISH frame and then reconnect. In the test case after reconnecting we first wait for the PUBLISH frame and then send the PUBREC frame, but changing the test to first send the PUBREC and then expect the (dup) PUBLISH works as well, see https://github.com/vernemq/vernemq/blob/master/apps/vmq_server/test/vmq_publish_SUITE.erl#L340-L342.

So in other words, I'm not able to reproduce this behaviour. Are you sure that the client connects with the same client-id?

Let me know if more information becomes available, then we'll reopen.

**Comment by srados** [comment]
Thanks, I will double check and let you know

**Comment by srados** [comment]
Hi again, I made a small Go mqtt client in order to demonstrate what happens
```
package main

import (
	"fmt"
	mqtt "github.com/eclipse/paho.mqtt.golang"
	"os"
	"os/signal"
	"syscall"
)
const Server = "tcp://xx.xx.xx.xx:1883"
const ClientID = "testclient"
const Topic = "sometopic"


func main() {
	opts := mqtt.NewClientOptions().AddBroker(Server)
	opts.SetClientID(ClientID)
	opts.SetAutoReconnect(false)
	opts.SetCleanSession(false)
	opts.SetProtocolVersion(4)
	opts.SetStore(mqtt.NewFileStore("./msgstore"))
	opts.SetConnectionLostHandler(makeConnLostHandler())
	mc := mqtt.NewClient(opts)
	token := mc.Connect()
	token.Wait()
	if token.Error() != nil {
		fmt.Printf("Failed to connect %s", token.Error())
		os.Exit(1)
	}
	mc.Subscribe(Topic, 2, makeReceiveHandler())
	fmt.Printf("subscribed to %s", Topic)
	errs := make(chan error, 2)
	go func() {
		c := make(chan os.Signal)
		signal.Notify(c, syscall.SIGINT)
		errs <- fmt.Errorf("%s", <-c)
	}()
	err := <-errs
	fmt.Printf("terminated %s", err)
}

func makeReceiveHandler() func(client mqtt.Client, msg mqtt.Message) {
	return func(client mqtt.Client, msg mqtt.Message) {
		fmt.Printf("got msg: %s\n",msg.Payload())
		//this should be uncommented when you want to break the client and make it leave i.xxx.msg in msgstore dir
		//os.Exit(1)
	}
}

func makeConnLostHandler() func(mqtt.Client,error) {
	return func(c mqtt.Client, e error) {
		fmt.Printf("conn lost: %s\n",e)
		os.Exit(1)
	}
}
```

These are the steps:

1. install vernemq like this `docker run -e "DOCKER_VERNEMQ_ALLOW_ANONYMOUS=on" -e "DOCKER_VERNEMQ_LOG__CONSOLE__LEVEL=debug" --name vernemq1 -p 1883:1883 -d erlio/docker-vernemq`

2. run the client code attached above 
3. do one mosquitto_pub publish like this `mosquitto_pub -m ttt -i someclient -q 2 -h xx.xx.xx.xx -t sometopic`
4. make sure you received the message by the Go client (if will print - got msg: ttt)
5. uncomment the os.Exit part here so the client breaks when it receives the message, thus leaving i.xxx.msg in msgstore dir
```
func makeReceiveHandler() func(client mqtt.Client, msg mqtt.Message) {
	return func(client mqtt.Client, msg mqtt.Message) {
		fmt.Printf("\ngot msg: %s",msg.Payload())
		//this should be uncommented when you want to break the client and make it leave i.xxx.msg in msgstore dir
		os.Exit(1)
	}
}
```
6. the Go client will break and leave i.xxx.msg file in the msgstore dir
7. comment the os.Exit(1) in makeReceiveHandler again, so the client runs in 'normal' mode
8. try to start the client
9. you should get this response (the client will exit)
```
subscribed to someticpi msg: ttt
conn lost: read tcp xx.xx.xx.xx:56002->xx.xx.xx.xx:1883: read: connection reset by peer

```
10. in vernemq logs you will see something like this
```
12:24:35.916 [debug] transition from offline --> online because of add_session
12:24:35.918 [debug] stopped connected session, due to unknown qos2 pubrec xxx
12:24:35.919 [debug] session normally stopped
12:24:35.919 [debug] transition from online --> offline because of 'DOWN'
```
where xxx is the number found in the i.xxx.msg

please note that this happens in most cases, but sometimes the client stays connected.

11. the i.xxx.msg staus in the msgstore

If you can't run this Go code, I can also provide binary that does the same.

**Comment by larshesel** [comment]
Thanks for the update - will reopen and take a look as soon as time permits.

**Comment by larshesel** [comment]
I've taken a look at this and even though I think retrying a PUBREL is bad behaviour in a client I think we should handle it. 

@dergraf, we could do like mosquitto and simply warn on this - though I think the root cause is that we're waiting to get the data from the queue when the client sends the PUBREL and the msgid is therefore not found. What do you think?

**Comment by larshesel** [comment]
The MQTT spec doesn't specifically disallow retrying PUBRECs, but it is explicit about that PUBLISH and PUBRELs MUST be retried after a reconnect. So in the case where the client resends the PUBREC, then it's doing something which isn't necessary as the broker MUST retry the original PUBLISH and then the client can respond with the PUBREC. So if I were the maintainer of the client I would consider this buggy behaviour. But perhaps I'm missing something ...

**Comment by dusanb94** [comment]
This is the tricky part of the client-side. I agree that there is no need to send PUBREC, but from the implementation side, it works in a protocol-compliant way: Receives PUBLISH message from the broker, stores it, and sends PUBPREC. In this particular scenario, the PUBPREC part is interrupted with os.Exit(1), so the client on the next running just continues where it stopped before - sending PUBREC without waiting for the broker to re-send the message with the same ID and DUP=1. As far as I know, this scenario is not addressed in the protocol spec. Anyways, even if the problem is on the client-side, I'm not sure that unknown PUBREC should disconnect the client since PUBREC means that the client _accepts ownership of the Application Message_(https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_QoS_2:_Exactly). Sending PUBREL for this PUBREC will not hurt because PUBREL means that the broker won't try to re-send the message. Since there is no such message (PUBREC is unknown), the broker won't try to send it anyway.
